encoding.go

```diff
@@ -0,0 +1,139 @@
+package viper
+
+import (
+	"github.com/spf13/viper/internal/encoding/dotenv"
+	"github.com/spf13/viper/internal/encoding/hcl"
+	"github.com/spf13/viper/internal/encoding/ini"
+	"github.com/spf13/viper/internal/encoding/javaproperties"
+	"github.com/spf13/viper/internal/encoding/json"
+	"github.com/spf13/viper/internal/encoding/toml"
+	"github.com/spf13/viper/internal/encoding/yaml"
+)
+
+// Encoder encodes Viper's internal data structures into a byte representation.
+// It's primarily used for encoding a map[string]any into a file format.
+type Encoder interface {
+	Encode(v map[string]any) ([]byte, error)
+}
+
+// Decoder decodes the contents of a byte slice into Viper's internal data structures.
+// It's primarily used for decoding contents of a file into a map[string]any.
+type Decoder interface {
+	Decode(b []byte, v map[string]any) error
+}
+
+// Codec combines [Encoder] and [Decoder] interfaces.
+type Codec interface {
+	Encoder
+	Decoder
+}
+
+type encodingError string
+
+func (e encodingError) Error() string {
+	return string(e)
+}
+
+const (
+	// ErrEncoderNotFound is returned when there is no encoder registered for a format.
+	ErrEncoderNotFound = encodingError("encoder not found for this format")
```

Comment:
I wonder if this should be an interface instead. That way, implementations wouldn't have to rely on Viper as a dependency.

Reply:
I wouldn't have an issue depending on Viper if I implemented an encoder for it?

Reply:
No, the encoder interface can be implemented implicitly.

Reply:
Of course. But isn‘t the error on the consumer side that would always be depending on Viper? Maybe I misunderstood the question, sorry for the noise.

Reply:
I'm talking about the _implementation_.